File: ios/SFR3/Models/Utils/ActionStatus.swift
//
//  ActionStatus.swift
//  SFR3
//
//  Created by Luiz SSB on 19/04/25.
//

enum ActionStatus<
    TResult: Equatable & Hashable,
    TError: Equatable & Hashable
>: Equatable, Hashable {
    case none,
         running,
         success(TResult),
         failure(TError)
}

File: web/src/models/utils/ActionStatus.ts
import { AppError } from "./AppError";

export type ActionStatus<TSuccess = undefined> =
  | { type: "none" | "running" }
  | ({ type: "success" } & (TSuccess extends undefined
      ? {
          result?: TSuccess;
        }
      : { result: TSuccess }))
  | {
      type: "failure";
      error: AppError;
    };
